File: frontend/src/utils/duplicateDetector.ts
/**
 * 重复产品检测工具
 *
 * 提供产品列表的重复判定键提取、重复分组、去重索引计算。
 *
 * 判定优先级：
 *   1) product_id（同一数据库产品）→ product_id:{id}
 *   2) customer_code + oe_number（无 product_id 的临时产品）→ code_oe:{customer_code}|{oe_number}
 *   3) detail_desc（产品名称）→ name:{detail_desc}
 */

export interface DuplicateGroup {
  /** 判定键 */
  key: string
  /** 展示文本（产品名称/编号） */
  display: string
  /** 在原数组中出现的索引列表 */
  indices: number[]
  /** 是否仅因与 existingKeys 冲突而标记为重复（订单内首次出现） */
  external: boolean
}

function normalize(value: any): string {
  if (value === null || value === undefined) return ''
  return String(value).trim()
}

/**
 * 为一行产品提取重复判定键和展示文本
 */
export function extractDuplicateKey(item: any): { key: string; display: string } {
  if (!item) return { key: '', display: '' }

  // 1) product_id 优先
  const productId = item.product_id
  if (productId !== null && productId !== undefined && productId !== 0) {
    const display =
      normalize(item.detail_desc) ||
      normalize(item.product_name) ||
      normalize(item.customer_code) ||
      normalize(item.oe_number) ||
      `产品#${productId}`
    return { key: `product_id:${productId}`, display }
  }

  // 2) customer_code + oe_number 兜底
  const code = normalize(item.customer_code)
  const oe = normalize(item.oe_number)
  if (code || oe) {
    return {
      key: `code_oe:${code}|${oe}`,
      display: code || oe,
    }
  }

  // 3) 产品名称兜底
  const name = normalize(item.detail_desc) || normalize(item.product_name)
  if (name) {
    return { key: `name:${name}`, display: name }
  }

  return { key: '', display: '' }
}

/**
 * 查找产品列表中的重复分组
 */
export function findDuplicates(
  items: any[],
  existingKeys: Set<string> = new Set()
): DuplicateGroup[] {
  const groups = new Map<string, DuplicateGroup>()

  items.forEach((item, idx) => {
    const { key, display } = extractDuplicateKey(item)
    if (!key) return
    if (!groups.has(key)) {
      groups.set(key, { key, display, indices: [], external: false })
    }
    groups.get(key)!.indices.push(idx)
  })

  const duplicates: DuplicateGroup[] = []
  groups.forEach((group) => {
    if (group.indices.length >= 2) {
      duplicates.push(group)
    } else if (existingKeys.has(group.key)) {
      group.external = true
      duplicates.push(group)
    }
  })

  duplicates.sort((a, b) =>
    (a.indices[0] || 0) - (b.indices[0] || 0)
  )
  return duplicates
}

/**
 * 计算"跳过重复行"后应保留的行索引
 */
export function filterDuplicateIndices(
  count: number,
  duplicateGroups: DuplicateGroup[]
): number[] {
  const skip = new Set<number>()
  for (const group of duplicateGroups) {
    if (!group.indices.length) continue
    if (group.external) {
      group.indices.forEach((i) => skip.add(i))
    } else {
      // 保留第一次出现，跳过后续
      group.indices.slice(1).forEach((i) => skip.add(i))
    }
  }
  const result: number[] = []
  for (let i = 0; i < count; i++) {
    if (!skip.has(i)) result.push(i)
  }
  return result
}

/**
 * 判断指定索引是否为重复行
 */
export function isDuplicateIndex(idx: number, groups: DuplicateGroup[]): boolean {
  return groups.some((g) => g.indices.includes(idx))
}

File: client/utils/duplicate_detector.py
"""
重复产品检测工具

提供预览行与订单项的重复判定键提取、重复分组、去重索引计算。
"""
from typing import Any, Dict, List, Optional, Set, Tuple


def _normalize(value: Any) -> str:
    """将值规范化为去除首尾空格的字符串，None 返回空字符串。"""
    if value is None:
        return ""
    return str(value).strip()


def extract_preview_duplicate_key(
    row: Any,
    headers: Optional[List[str]] = None,
    model_col_idx: Optional[int] = None,
) -> Tuple[str, str]:
    """
    为预览表中的一行提取重复判定键和展示文本。

    返回 (key, display)。key 为空字符串表示无法判定（不参与重复统计）。
    """
    if isinstance(row, dict):
        product_id = row.get('product_id')
        if product_id is not None and _normalize(product_id):
            display = _normalize(row.get('customer_code') or row.get('oe_number') or product_id)
            return (f"product_id:{product_id}", display)

        code = _normalize(row.get('customer_code', ''))
        oe = _normalize(row.get('oe_number', ''))
        if code or oe:
            display = code or oe
            return (f"code_oe:{code}|{oe}", display)
        return ("", "")

    # Excel 原始行（list / tuple），按 Model/客户产品编号列判定
    if isinstance(row, (list, tuple)) and model_col_idx is not None:
        if 0 <= model_col_idx < len(row):
            model = _normalize(row[model_col_idx])
            if model:
                return (f"model:{model}", model)
    return ("", "")


def find_preview_duplicates(
    rows: List[Any],
    headers: Optional[List[str]] = None,
    model_col_idx: Optional[int] = None,
    existing_keys: Optional[Set[str]] = None,
) -> List[Dict[str, Any]]:
    """
    查找预览行中的重复分组。

    Args:
        rows: 预览行列表，元素可以是 dict（手动新增）或 list（Excel 原始行）。
        headers: Excel 表头，用于定位 Model 列（可选）。
        model_col_idx: Model 列在 Excel 行中的索引（可选）。
        existing_keys: 已存在产品的判定键集合（如当前订单已有产品、客户-产品表已有 Model）。

    Returns:
        重复分组列表，每个分组包含：
        - key: 判定键
        - display: 展示文本
        - indices: 在 rows 中出现的索引列表
        - external: 是否仅因与 existing_keys 冲突而标记为重复
    """
    existing_keys = existing_keys or set()
    groups: Dict[str, Dict[str, Any]] = {}

    for idx, row in enumerate(rows):
        key, display = extract_preview_duplicate_key(row, headers, model_col_idx)
        if not key:
            continue
        if key not in groups:
            groups[key] = {
                "key": key,
                "display": display,
                "indices": [],
                "external": False,
            }
        groups[key]["indices"].append(idx)

    duplicates: List[Dict[str, Any]] = []
    for key, group in groups.items():
        if len(group["indices"]) >= 2:
            duplicates.append(group)
        elif key in existing_keys:
            group["external"] = True
            duplicates.append(group)

    duplicates.sort(key=lambda g: g["indices"][0] if g["indices"] else 0)
    return duplicates


def filter_duplicate_indices(
    row_count: int,
    duplicate_groups: List[Dict[str, Any]],
) -> List[int]:
    """
    计算选择"跳过重复行"后应保留的行索引。

    - 内部重复：保留第一次出现，跳过后续。
    - 与外部已存在产品冲突：跳过所有冲突行。
    """
    skip: Set[int] = set()
    for group in duplicate_groups:
        indices = group.get("indices", [])
        if not indices:
            continue
        if group.get("external"):
            skip.update(indices)
        else:
            skip.update(indices[1:])
    return [i for i in range(row_count) if i not in skip]
